lib/auth/init.go

```diff
@@ -483,6 +482,49 @@ func Init(cfg InitConfig, opts ...ServerOption) (*Server, error) {
 	return asrv, nil
 }
 
+func setAuthPreference(asrv *Server, newAuthPref services.AuthPreference) error {
+	storedAuthPref, err := asrv.GetAuthPreference()
+	if err != nil {
+		if !trace.IsNotFound(err) {
+			return trace.Wrap(err)
+		}
+		storedAuthPref = nil
+	}
+	shouldReplace, err := shouldInitReplaceResource(storedAuthPref, newAuthPref)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	if shouldReplace {
+		err = asrv.SetAuthPreference(newAuthPref)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		log.Infof("Updating auth preference: %v.", newAuthPref)
+	}
+	return nil
+}
+
+// shouldInitReplaceResource determines whether the candidate resource should
+// be used to replace the stored resource during auth server initialization.
+//
+// For more details, see Section "Auth server initialization"
+// in rfd/0016-dynamic-configuration.md
+func shouldInitReplaceResource(stored, candidate types.ResourceWithOrigin) (bool, error) {
+	if candidate.Origin() != types.OriginConfigFile && candidate.Origin() != types.OriginDefaults {
+		return false, trace.BadParameter("candidate origin must be either config-file or defaults (this is a bug)")
+	}
+	if stored == nil || stored.Origin() == types.OriginConfigFile || stored.Origin() == types.OriginDefaults {
+		return true, nil
+	}
+	if candidate.Origin() == types.OriginConfigFile {
+		// Log a warning when about to overwrite a dynamically configured resource.
+		log.Warnf("Stored %v resource with origin %q is to be discarded in favor of resource with origin %q.",
+			stored.GetKind(), stored.Origin(), candidate.Origin())
+		return true, nil
+	}
```

Comment:
this should `return false, nil`
as a Cloud user, I'd want to call `tctl create mydynamicpreference.yaml` and not have it wiped out when the auth server restarts next time

unless I'm misunderstanding the design, this whole function should be `return candidate.Origin() == types.OriginConfigFile || candidate.Origin() == types.OriginDefaults`

Reply:
Your suggestion would replace any stored dynamic configuration with the default one upon restart.

Reply:
NVM, I mis-undestood the RFD